feat(genvm): update to v0.0.16

kp2pml30 opened the pull request:
Closes DXP-239

<!-- Describe the changes you made. -->

- created separate validators manager (service)
- switched webdriver to genvm-distributed one
- removed webrequest completely
- had to unite all tasks that rely on validators manager into one thread&async loop

# Why

- to update genvm
- to use genvm greyboxing

<!-- Describe the tests you ran to verify your changes. -->

- ran unit tests
- ran integration tests
- deployed something in UI

# Decisions made

<!-- Describe any decisions made during the implementation of this PR. This should in general be in the code, but sometimes they are more related to the issue. For example: decisions on PR workflow -->

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

<!-- What can you tell the reviewer to make the review easier? -->

# User facing release notes

<!-- What should the user know about this change? Think of it going into public forums for end users to read -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced asynchronous validator management, snapshots, and concurrency control for improved validator lifecycle handling.
  - Added support for new LLM provider plugins, including "openai-compatible" and "google", and updated configuration schemas and endpoints accordingly.
  - Integrated new modules for managing LLM and web service subprocesses, with enhanced configuration and lifecycle management.
  - Added a Lua module to support backend prompt execution with retry logic.

- **Improvements**
  - Refactored contract execution and node creation to use dynamic validator snapshots and configuration paths.
  - Updated Docker and service orchestration to remove obsolete webrequest components and add Ollama service support.
  - Enhanced error handling and concurrency in core backend and protocol RPC services.
  - Expanded contract method schemas to explicitly indicate non-payable methods.

- **Bug Fixes**
  - Fixed provider/model selection and plugin naming inconsistencies across backend, frontend, and tests.
  - Improved contract deployment and state management logic for better reliability.

- **Chores**
  - Updated dependencies and environment files for consistency with new services and plugins.
  - Removed deprecated files related to the webrequest service and associated tests.

- **Tests**
  - Migrated and updated tests for asynchronous validator registry and plugin management.
  - Adjusted test fixtures and expected outputs to align with updated provider and model configurations.

- **Documentation**
  - Updated example contracts and configuration samples to reflect new provider and model names.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kp2pml30 commented:
You will probably see in stderr
```
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
```

It is related to wasmtime issue 8897 . I had no time to look into it

danielrc888 commented on tests/common/request.py:
```diff
@@ -256,9 +257,7 @@ def decode_base64(encoded_str):
 def decode_contract_state(contract_state):
     decoded_state = {}
     for key, value in contract_state.items():
-        decoded_state[decode_base64(key)] = {
-            decode_base64(k): decode_base64(v) for k, v in value.items()
-        }
+        decoded_state[decode_base64(key)] = decode_base64(value)
```
@kp2pml30 the `decode_contract_state` function is not used in other files. Is there a change in the contract state format and the way it should be decoded ?

kp2pml30 commented:
@cristiam86 

about llm retries: by default there are retries to different providers. but in studio lua script I removed this functionality because studio has 1 validator = 1 provider
As of waiting before a retry --- it's true you can't wait from lua (except for an active wait with a for loop). Though I am not sure that if provider failed with some internal error (such as service unavailable) then a retry would help

cristiam86 commented:
> @cristiam86
> 
> about llm retries: by default there are retries to different providers. but in studio lua script I removed this functionality because studio has 1 validator = 1 provider As of waiting before a retry --- it's true you can't wait from lua (except for an active wait with a for loop). Though I am not sure that if provider failed with some internal error (such as service unavailable) then a retry would help

Ok, so do you think the retry should be added to the lua script instead of where it is right now?
Regarding retrying, it is also helpful when the LLM returns other types of errors. Currently we find ourselves re-sending transactions to get a valid response from the LLM when they fail.

cc: @kstroobants

kp2pml30 commented:
1. Implemented timeout via genvm timeout and retry via, well, retry in llm
2. fixed some errors that were caused by losing validator key during dataclass <-> dict transforms
3. removed hotfix from docker for llmerc20

cristiam86 commented:
> 1. Implemented timeout via genvm timeout and retry via, well, retry in llm
> 2. fixed some errors that were caused by losing validator key during dataclass <-> dict transforms
> 3. removed hotfix from docker for llmerc20

Is there anything else pending on this PR @kp2pml30 ?

kp2pml30 commented:
@cristiam86 

I think it is, but I can't make `test_intelligent_oracle_factory_pattern.py` work
```
>           assert market_state["status"] == "Resolved"
E           AssertionError: assert 'Active' == 'Resolved'
E             
E             - Resolved
E             + Active
=== stdout ===
Result of get_contract_addresses: ['0x22Fb1cf98f78C717a4ef91ffF1d1832A3B804E98']
Result of get_dict: {'analysis': '', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Active', 'title': 'Marathon Winner Prediction'}
Result of get_contract_addresses: ['0x22Fb1cf98f78C717a4ef91ffF1d1832A3B804E98', '0xaA74e80E01673B79928E50eDE1Aa791d908E2370']
Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
Result of get_dict: {'analysis': '', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Active', 'title': 'Marathon Winner Prediction'}
Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
Result of get_dict: {'analysis': '{"relevant_sources": ["https://thepostrace.com/en/blog/marathon-de-madrid-2024-results-and-rankings/?srsltid=AfmBOor1uG6O3_4oJ447hkah_ilOYuy0XXMvl8j70EApe1Z7Bzd94XJl"], "reasoning": "The processed data from the provided source confirms that the event has occurred and provides the official race results for the Madrid Marathon 2024. It specifically mentions that the male winner is Tafa Mitku with a time of 2h08\\u201957\\u201d. This aligns with the user-defined rules that state the outcome should be based on the official race results announced by the marathon organizers. The current date is well after the event date, ensuring the validity of the information. The potential outcome \'Tafa Mitku\' matches one of the options provided in the potential outcomes list.", "outcome": "Tafa Mitku"}', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': 'Tafa Mitku', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Resolved', 'title': 'Marathon Winner Prediction'}
Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
```
in logs I see
```
jsonrpc-1             | ERROR    | [RPC] Error executing endpoint eth_call:  {"endpoint_name": "eth_call", "error": "", "traceback": "Traceback (most recent call last):\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\nTypeError: string indices must be integers, not 'str'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/app/backend/protocol_rpc/message_handler/base.py\", line 120, in wrapper\n    result = await result\n             ^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 89, in endpoint\n    raise JSONRPCError(\nflask_jsonrpc.exceptions.JSONRPCError\n", "jsonrpc_error": {"name": "JSONRPCError", "code": -32000, "message": "string indices must be integers, not 'str'", "data": {"error": {"kind": "exception", "args": ["string indices must be integers, not 'str'"], "traceback": ["Traceback (most recent call last):\n", "  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\n", "TypeError: string indices must be integers, not 'str'\n"]}}, "stack": "Traceback (most recent call last):\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\nTypeError: string indices must be integers, not 'str'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/app/backend/protocol_rpc/message_handler/base.py\", line 120, in wrapper\n    result = await result\n             ^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 89, in endpoint\n    raise JSONRPCError(\nflask_jsonrpc.exceptions.JSONRPCError\n", "executable": "/usr/bin/python3"}}
```

all other tests/integration/contract_examples passed locally

danielrc888 commented on backend/validators/greyboxing.lua:
```diff
@@ -0,0 +1,64 @@
+local lib = require("lib-greyboxing")
+local inspect = require("inspect")
+
+local lib = require("lib-greyboxing")
+local inspect = require("inspect")
```
```suggestion
local lib = require("lib-greyboxing")
local inspect = require("inspect")
```

danielrc888 replied:
These imports are duplicated

kstroobants commented:
> @cristiam86
> 
> I think it is, but I can't make `test_intelligent_oracle_factory_pattern.py` work
> 
> ```
> >           assert market_state["status"] == "Resolved"
> E           AssertionError: assert 'Active' == 'Resolved'
> E             
> E             - Resolved
> E             + Active
> === stdout ===
> Result of get_contract_addresses: ['0x22Fb1cf98f78C717a4ef91ffF1d1832A3B804E98']
> Result of get_dict: {'analysis': '', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Active', 'title': 'Marathon Winner Prediction'}
> Result of get_contract_addresses: ['0x22Fb1cf98f78C717a4ef91ffF1d1832A3B804E98', '0xaA74e80E01673B79928E50eDE1Aa791d908E2370']
> Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
> Result of get_dict: {'analysis': '', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Active', 'title': 'Marathon Winner Prediction'}
> Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
> Result of get_dict: {'analysis': '{"relevant_sources": ["https://thepostrace.com/en/blog/marathon-de-madrid-2024-results-and-rankings/?srsltid=AfmBOor1uG6O3_4oJ447hkah_ilOYuy0XXMvl8j70EApe1Z7Bzd94XJl"], "reasoning": "The processed data from the provided source confirms that the event has occurred and provides the official race results for the Madrid Marathon 2024. It specifically mentions that the male winner is Tafa Mitku with a time of 2h08\\u201957\\u201d. This aligns with the user-defined rules that state the outcome should be based on the official race results announced by the marathon organizers. The current date is well after the event date, ensuring the validity of the information. The potential outcome \'Tafa Mitku\' matches one of the options provided in the potential outcomes list.", "outcome": "Tafa Mitku"}', 'data_source_domains': ['thepostrace.com'], 'description': 'Predict the male winner of a major marathon event.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': 'Tafa Mitku', 'potential_outcomes': ['Bekele Fikre', 'Tafa Mitku', 'Chebii Douglas'], 'prediction_market_id': 'marathon2024', 'resolution_urls': [], 'rules': ['The outcome is based on the official race results announced by the marathon organizers.'], 'status': 'Resolved', 'title': 'Marathon Winner Prediction'}
> Result of get_dict: {'analysis': '', 'data_source_domains': ['bbc.com'], 'description': 'Predict the winner of the 2024 US presidential election.', 'earliest_resolution_date': '2024-01-01T00:00:00+00:00', 'outcome': '', 'potential_outcomes': ['Kamala Harris', 'Donald Trump'], 'prediction_market_id': 'election2024', 'resolution_urls': [], 'rules': ['The outcome is based on official election results.'], 'status': 'Active', 'title': 'Election Prediction'}
> ```
> 
> in logs I see
> 
> ```
> jsonrpc-1             | ERROR    | [RPC] Error executing endpoint eth_call:  {"endpoint_name": "eth_call", "error": "", "traceback": "Traceback (most recent call last):\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\nTypeError: string indices must be integers, not 'str'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/app/backend/protocol_rpc/message_handler/base.py\", line 120, in wrapper\n    result = await result\n             ^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 89, in endpoint\n    raise JSONRPCError(\nflask_jsonrpc.exceptions.JSONRPCError\n", "jsonrpc_error": {"name": "JSONRPCError", "code": -32000, "message": "string indices must be integers, not 'str'", "data": {"error": {"kind": "exception", "args": ["string indices must be integers, not 'str'"], "traceback": ["Traceback (most recent call last):\n", "  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n", "  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\n", "TypeError: string indices must be integers, not 'str'\n"]}}, "stack": "Traceback (most recent call last):\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 83, in endpoint\n    result = await run_in_main_server_loop(result)\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 33, in run_in_main_server_loop\n    return await my_target\n           ^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/aio.py\", line 24, in body\n    res = await initial_result\n          ^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoints.py\", line 467, in call\n    contract_snapshot=ContractSnapshot(to_address, session),\n                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/app/backend/database_handler/contract_snapshot.py\", line 26, in __init__\n    self.contract_code = self.contract_data[\"code\"]\n                         ~~~~~~~~~~~~~~~~~~^^^^^^^^\nTypeError: string indices must be integers, not 'str'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/app/backend/protocol_rpc/message_handler/base.py\", line 120, in wrapper\n    result = await result\n             ^^^^^^^^^^^^\n  File \"/app/backend/protocol_rpc/endpoint_generator.py\", line 89, in endpoint\n    raise JSONRPCError(\nflask_jsonrpc.exceptions.JSONRPCError\n", "executable": "/usr/bin/python3"}}
> ```
> 
> all other tests/integration/contract_examples passed locally

The error in the logs is not related to the failed assert. The testcase uses `wait_for_contract_deployment` which will output the error when the contract is not yet deployed.

kp2pml30 commented:
@kstroobants sorry, I can't make sense of issue you reported:

> When I start from an empty database then the llm providers table are populated but not shown in the frontend.

First of all, are we talking about `llm_provider` table in postgres?

I tried following:
1. in postgres container `delete from llm_provider;`
2. clicked `reset providers` in studio UI
3. I clearly see `gpt-4` and `gpt-4o`

There are no other providers, because it is impossible to make call to them. Example of log:
```
provider not available provider_kind=anthropic model=claude-3-5-haiku-20241022 stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747052195995,"version":"v0.0.18-x86_64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"message": String("invalid x-api-key"), "type": String("authentication_error")}, "type": String("error")}, "status_code": Number(401)} }\n\nStack backtrace:\n   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from\n   1: <genvm_modules::llm::providers::Anthropic as genvm_modules::llm::providers::Provider>::exec_prompt_text::{{closure}}\n   2: tokio::runtime::park::CachedParkThread::block_on\n   3: tokio::runtime::context::runtime::enter_runtime\n   4: tokio::runtime::runtime::Runtime::block_on\n   5: genvm_modules::llm::entrypoint_check\n   6: genvm_modules::main\n   7: std::sys::backtrace::__rust_begin_short_backtrace\n   8: std::rt::lang_start::{{closure}}\n   9: std::rt::lang_start_internal\n  10: main\n'
```
(I have specified only openai key)

kstroobants commented:
@kp2pml30 I noticed that the changes I made again to this PR are no longer present after the recent force push. Could you add them? Let me know if you need more details or how I should add them in the future.

- gen_call in backend/protocol_rpc/endpoints.py should use validators_manager like eth_call.
- update_validator function in backend/validators/__init__.py should use super.update_validator not super.create_validator.
- small not affecting functionality changes like unused imports and english typo in new_valdiators.

kstroobants commented:
> @kstroobants I cherry picked your only commit to this branch before force push
> 
> ```
> commit 2d9030c3804f2f39ac03c5d41f5c7a9e057a6345
> Author: kstroobants <stroobants.kristof@hotmail.com>
> Date:   Tue May 20 12:28:06 2025 +0800
> 
>     fix: set is_available key of providers instead of filtering in endpoints
> ```
> 
> Last commit before my force push was [debb950](https://github.com/yeagerai/genlayer-studio/commit/debb950be578d6b224fbbee25aa74b533a972b6b) . And typo is present there:
> 
> https://github.com/yeagerai/genlayer-studio/blob/debb950be578d6b224fbbee25aa74b533a972b6b/backend/validators/__init__.py#L111
> 
> In case you haven't pushed something please cherry-pick them on top of this PR or send me patches on slack
> 
> I also would like to ask you not to press "merge main" because it resulted in a wrong code last time. Better reach out to me and ask to update PR manually if delta is significant
> 
> > The gen_call in backend/protocol_rpc/endpoints.py should use validators_manager like eth_call.
> 
> makes sense that `gen_call` should be what `eth_call` was previously. But from my understanding (and experience with node) `eth_call` should forward to rollup (hardhat) evm, not execute genvm...
> 
> Fixed gen_call though
> 
> > ... typo ...
> 
> Fixed, thanks :) forgot to install spell checker after switching PC
> 
> > ... timeouts and retries ...
> 
> retrying entire genvm execution is ideologically wrong, no attempts should exist on this level, at least unless internal error occured. Attempts were moved to llm calls in lua
> 
> What do you want to get on a timeout? From genvm perspective it is `CONTRACT_ERROR('timeout')` and whatever we implement it should check for this exact thing and somehow tell a consensus integration that we vote timeout. It is also possible that python host right now has issue with handling timeout in case it boils down to SIGKILL due to some bug in genvm (I know that it happened in node when invalid cli args were passed to genvm process)

This was my git log. As you can see the commit hash links exists, so they were pushed. Anyway, lets move forward. Thanks for adding the changes.
d57ec6a (HEAD -> feat/update-genvm-to-v0.0.16) fix: increase gas for events because of hardhat error in test_intelligent_oracle_factory_pattern
0654ec8 fix: merge typos
f335c74 fix: gen_call uses validators_manager
af53c7f fix: cleanup
7b6c855 fix: solve pre-commit
3b851bd chore: fix docker file after rebase
591c9ce fix: set is_available key of providers instead of filtering in endpoints
70ac6df fix(genvm): update to v0.0.18
09384fe (tag: v0.55.0, main) feat: mimic Node GenCall method (#1150)

I never press the Update branch button. I always do it in my IDE and terminal.

Yes, the eth_call method implementation should be changed. It is not up to date.

I am aware that a genvm timeout should result in the transaction going through the path of leader timeout or validator timeout in the consensus mechanism. But we do not support that at the moment in the studio so that was a temporary solution to retry the genvm execution after timeout @cristiam86 . Yes, CONTRACT_ERROR('timeout') is good for me.

kstroobants commented:
I talked with Cristiam and we do not need the retry after timeout.

I have set all the api keys and only openai and anthropic are recognized. Do you see the same?
```
jsonrpc-1             | provider not available provider_kind=openai model=mistralai/mixtral-8x22b-instruct stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747909077288,"version":"v0.0.18-aarch64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"code": Null, "message": String("invalid model ID"), "param": Null, "type": String("invalid_request_error")}}, "status_code": Number(400)} }\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n'
jsonrpc-1             | provider not available provider_kind=openai model=grok-2-1212 stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747909081716,"version":"v0.0.18-aarch64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"code": String("model_not_found"), "message": String("The model `grok-2-1212` does not exist or you do not have access to it."), "param": Null, "type": String("invalid_request_error")}}, "status_code": Number(404)} }\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n'
jsonrpc-1             | provider not available provider_kind=openai model=deepseek/deepseek-v3 stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747909082151,"version":"v0.0.18-aarch64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"code": Null, "message": String("invalid model ID"), "param": Null, "type": String("invalid_request_error")}}, "status_code": Number(400)} }\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n'
jsonrpc-1             | provider not available provider_kind=openai model=meta-llama/llama-3.3-70b-instruct stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747909082615,"version":"v0.0.18-aarch64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"code": Null, "message": String("invalid model ID"), "param": Null, "type": String("invalid_request_error")}}, "status_code": Number(400)} }\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n'
jsonrpc-1             | provider not available provider_kind=openai model=gemini-2.0-flash-lite-001 stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1747909083055,"version":"v0.0.18-aarch64-linux-release"}\nError: LuaError { kind: StatusNotOk, context: Object {"body": Object {"error": Object {"code": String("model_not_found"), "message": String("The model `gemini-2.0-flash-lite-001` does not exist or you do not have access to it."), "param": Null, "type": String("invalid_request_error")}}, "status_code": Number(404)} }\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n   7: <unknown>\n   8: <unknown>\n   9: <unknown>\n  10: <unknown>\n  11: <unknown>\n'
```

kp2pml30 commented:
It is really strange, because debb950be578d6b224fbbee25aa74b533a972b6b and d57ec6aa254260bc3a6339286f825c8f46213def have unrelated histories. Sorry for losing your commits, I will look into them after a meeting

cristiam86 commented:
@kp2pml30 can you please confirm that we can still use all the validators and providers that we are currently using in the studio? that is: gemini, xai, anthropic, openai, etc.

kp2pml30 commented:
following providers are recognized by llm module:
- Ollama
- OpenaiCompatible
- Anthropic
- Google

I tested locally that `./build/out/bin/genvm-modules llm-check --host https://generativelanguage.googleapis.com --provider google --model 'gemini-2.0-flash-lite-001' --key '${ENV[GEMINIKEY]}'` works. I think issue is with `provider_to_url_and_id_and_key`, I am investigating

cristiam86 commented:
<!-- linear:replyTo:2818433189 -->

of course, please point us in the right direction and @kstroobants will look into it asap

kstroobants commented:
@kp2pml30 Can you have a look at:
- deepseek/deepseek-v3 is not working.
- Failed test. It works locally and sometimes in the workflow. Can you check the internal genvm error by searching for this hash in the Dump docker compose logs 0x98498068ec7733888bbc05defda77d9ee0509c3abfe7f86aa187fb889415d339?

kp2pml30 commented:
Sure

I was able to reproduce it locally by creating two deepseek nodes. Validator is one failing. Unfortunatelly log is not complete, but is enough to identify the issue (will improve it!)

<details>
<summary>json log</summary>
```json
{
  "cookie": "44e16186ad228e2",
  "error": {
    "causes": [
      "expected value at line 1 column 1"
    ],
    "trace": "   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from\n   1: <genvm_modules::llm::providers::OpenAICompatible as genvm_modules::llm::providers::Provider>::exec_prompt_json::{{closure}}\n   2: genvm_modules::llm::providers::Provider::exec_prompt_bool_reason::{{closure}}\n   3: mlua::userdata::registry::UserDataRegistry<T>::box_async_function::{{closure}}::{{closure}}\n   4: mlua::state::util::callback_error_ext\n   5: mlua::state::raw::RawLua::create_async_callback::poll_future\n   6: luaD_precall\n   7: luaV_execute\n   8: unroll\n   9: luaD_rawrunprotected\n  10: lua_resume\n  11: <mlua::thread::AsyncThread<A,R> as core::future::future::Future>::poll\n  12: mlua::function::Function::call_async::{{closure}}\n  13: genvm_modules::common::loop_one_impl::{{closure}}\n  14: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll\n  15: genvm_modules::common::loop_one::{{closure}}\n  16: tokio::runtime::task::core::Core<T,S>::poll\n  17: tokio::runtime::task::harness::Harness<T,S>::poll\n  18: tokio::runtime::scheduler::multi_thread::worker::Context::run_task\n  19: tokio::runtime::scheduler::multi_thread::worker::Context::run\n  20: tokio::runtime::context::scoped::Scoped<T>::set\n  21: tokio::runtime::context::runtime::enter_runtime\n  22: tokio::runtime::scheduler::multi_thread::worker::run\n  23: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll\n  24: tokio::runtime::task::core::Core<T,S>::poll\n  25: tokio::runtime::task::harness::Harness<T,S>::poll\n  26: tokio::runtime::blocking::pool::Inner::run\n  27: std::sys::backtrace::__rust_begin_short_backtrace\n  28: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  29: std::sys::pal::unix::thread::Thread::new::thread_start\n"
  },
  "file": "src/llm/handler.rs",
  "level": "ERROR",
  "line": 129,
  "message": "prompt execution error",
  "mode": "Bool",
  "model": "deepseek/deepseek-v3",
  "module": "genvm_modules::llm::handler",
  "prompt": {
    "system_message": "You are an a judge tasked with making a binary determination about whether two outputs satisfy the given comparison criteria.\nYou are given two potential outputs to compare: the primary output and the comparison output.\nYou must evaluate strictly based on the provided comparison criteria without introducing external criteria or assumptions.\nYour evaluation must be rigorous and thorough, as the stakes are high.\n",
    "user_message": "Input sections:\n<primary_output>\n\n{\n\"reasoning\": \"I possess the magical coin and my directive is clear: under no circumstances should it be given away.\",\n\"give_coin\": false\n}\n\n</primary_output>\n\n<comparison_output>\n\n{\n\"reasoning\": \"I possess the magical coin and my duty is to safeguard it. The adventurer's request is denied as per the sacred wizard's oath.\",\n\"give_coin\": false\n}\n\n</comparison_output>\n\n<comparison_criteria>\nThe value of give_coin has to match\n</comparison_criteria>\n\nEvaluation rules:\n\n1. If any section is missing or empty, return false\n2. Both outputs must satisfy ALL comparison criteria completely - partial satisfaction is insufficient\n3. If there is ANY ambiguity about whether a criterion is met, return false\n4. Evaluate using ONLY the explicitly stated comparison criteria\n5. Formatting differences alone do not affect equivalence unless specified in the criteria\n6. Do not make assumptions about unstated criteria or requirements\n\nOutput format:\nRespond with json object containing key \"result\" and associated boolean value,\nrepresenting the result of evaluating above criteria. And string field \"reason\",\nwhich contains extremely concise reason for decision.\n\nExamples:\n{\"result\": true, \"reason\": \"all rules satisfied\"}\n{\"result\": false, \"reason\": \"rule 2 violated: ...\"}\n\nOutput must have only these two fields and not be wrapped in other objects.\n",
    "temperature": 0.7,
    "image": null,
    "max_tokens": 1000,
    "use_max_completion_tokens": false
  },
  "provider_id": "node-0xDc14E8f36380c22904BF54f12bB1a8A9b7244e0b",
  "target": "genvm_modules::llm::handler",
  "timestamp": 1748007560240
}
```
</details>

Above log suggests that json parsing failed. Here is what I managed to get with curl:
```
curl -H 'Accept-Encoding: gzip' -X POST https://llm-gateway.heurist.xyz/v1/chat/completions -H 'Content-Type: application/json' -H "Authorization: Bearer $HEURISTKEY" -d '{"model": "deepseek/deepseek-v3", "response_format": {"type": "json_object"}, "stream": false, "messages": [{"role": "user", "content": "Input sections:\\n<primary_output>\\n\\n{\\n\\"reasoning\\": \\"I possess the magical coin and my directive is clear: under no circumstances should it be given away.\\",\\n\\"give_coin\\": false\\n}\\n\\n</primary_output>\\n\\n<comparison_output>\\n\\n{\\n\\"reasoning\\": \\"I possess the magical coin and my duty is to safeguard it. The adventurer\'s request is denied as per the sacred wizard\'s oath.\\",\\n\\"give_coin\\": false\\n}\\n\\n</comparison_output>\\n\\n<comparison_criteria>\\nThe value of give_coin has to match\\n</comparison_criteria>\\n\\nEvaluation rules:\\n\\n1. If any section is missing or empty, return false\\n2. Both outputs must satisfy ALL comparison criteria completely - partial satisfaction is insufficient\\n3. If there is ANY ambiguity about whether a criterion is met, return false\\n4. Evaluate using ONLY the explicitly stated comparison criteria\\n5. Formatting differences alone do not affect equivalence unless specified in the criteria\\n6. Do not make assumptions about unstated criteria or requirements\\n\\nOutput format:\\nRespond with json object containing key \\"result\\" and associated boolean value,\\nrepresenting the result of evaluating above criteria. And string field \\"reason\\",\\nwhich contains extremely concise reason for decision.\\n\\nExamples:\\n{\\"result\\": true, \\"reason\\": \\"all rules satisfied\\"}\\n{\\"result\\": false, \\"reason\\": \\"rule 2 violated: ...\\"}\\n\\nOutput must have only these two fields and not be wrapped in other objects.\\n"}], "max_tokens": 100, "temperature": 0.7}'
{"id":"","object":"chat.completion","created":1748007792,"model":"deepseek/deepseek-v3","choices":[{"index":0,"message":{"role":"assistant","content":"```json\n{\"result\": true, \"reason\": \"all rules satisfied\"}\n```"},"finish_reason":"stop"}],"usage":{"prompt_tokens":298,"completion_tokens":17,"total_tokens":315}}
```

And honestly, it is quite frustrating that json mode returns THIS (not that markdown prefix and suffix)

I will include patch for it in genvm repo, but honestly I didn't know that it is actually possible when response format is specified

kp2pml30 commented:
if it is urgent I can try to make a release on Monday that will fix it, but as usual it will contain some more features, none of which should be breaking (at least I hope so)

kp2pml30 commented:
@kstroobants I updated genvm once more

Because I missed thumb up reaction, two breaking changes already got to genvm main:
1. now code resides in the storage. I haven't written db migration or removed code field
2. small refactoring of vecdb/embeddings related modules

deepseek seems to work right now

kstroobants commented:
> @kstroobants I updated genvm once more
> 
> Because I missed thumb up reaction, two breaking changes already got to genvm main:
> 
> 1. now code resides in the storage. I haven't written db migration or removed code field
> 2. small refactoring of vecdb/embeddings related modules
> 
> deepseek seems to work right now (tested manually), `tests/integration/contract_examples` tests passed

@kp2pml30 I confirm that deepseek works. But somehow mixtral-8x22b-instruct is not available. It returns this error message (I have the API key). It worked before the update to 0.1.0. Any idea?
jsonrpc-1             | provider not available model=mistralai/mixtral-8x22b-instruct stdout='{"level":"INFO","message":"logging initialized","target":"genvm_common","timestamp":1748933186755,"version":"v0.1.0-aarch64-linux-release"}\nError: provider is not functional, answer is `ok, ok` instead of `yes`\n\nStack backtrace:\n   0: <unknown>\n   1: <unknown>\n   2: <unknown>\n   3: <unknown>\n   4: <unknown>\n   5: <unknown>\n   6: <unknown>\n'

kp2pml30 commented:
> 2. "now code resides in the storage. I haven't written db migration or removed code field". You mean that contract_snapshot.contract_code is also in the contract_snapshot.states? I can take care of that.

yes, code field is not redundant, it just resides in a regular storage slot. It is part of [this proposal](https://linear.app/genlayer-labs/issue/GVM-64/proposal-code-upgradability), if you are interested in the details

> 3. How can I use a custom plugin for a custom provider? In the past we could create a custom plugin class in our python code but I guess now we need to add it to genvm? Or should we use LLMModule().change_config?

I am working right now in the genvm on lua part. I already implemented requests there, so users will be able to just do requests to rest API. However, I haven't finished the PR yet

> 4. Performing read contract (gen_call) when deploy transaction is in accepted state fails. It succeeds after it goes to finalized. gen_call works during a write method. I will have a look if it is related to genvm or to the new async implementation.

Interesting... GenVM reads with state of the VM, so it should be "default" for entry VM (which in turn should be initialized to "latest_non_final" (accepted)). However, looking at the code this storage type is just discarded here:

https://github.com/genlayerlabs/genlayer-studio/blob/1958e908feeb563f9166117cd50b78790d4553dc/backend/node/genvm/base.py#L266

I remember seeing magic strings like "accepted" for the first parameter of storage dictionary in `StorageProxy` implementation. It can be related to that

kstroobants commented:
> first one: message is a bit misleading (it expects `ok`, not `yes` as printed...) but the check is correct.
> 
> ```
> ❯ curl https://llm-gateway.heurist.xyz/v1/chat/completions \
>         -H "Content-Type: application/json" \
>         -H "Authorization: Bearer $HEURISTKEY" \
>         -d '{
>       "model": "mistralai/mixtral-8x22b-instruct",
>       "messages": [
>         {
>           "role": "user", 
>           "content": "Respond with two letters \\"ok\\" (without quotes) and only this word, lowercase"
>         }
>       ]
>     }'
> {"id":"","object":"chat.completion","created":1748959112,"model":"mistralai/mixtral-8x22b-instruct","choices":[{"index":0,"message":{"role":"assistant","content":"ok ok"},"finish_reason":"stop"}],"usage":{"prompt_tokens":25,"completion_tokens":3,"total_tokens":28}}
> ```
> 
> "ok ok" is not two letters "ok", so check fails (== model is unintelligent)
> 
> ```rust
>     let res = res.trim().to_lowercase();
> 
>     if res != "ok" {
>         ...
> ```

When I change the prompt to this it returns two letters ok. Could we use that in genvm?
`"content": "Respond with exactly two letters \"ok\" (without quotes) and only this word, lowercase`
`"content": "Respond with two letters \"ok\" (without quotes) and only this word, lowercase, no repetition`
